No processing of transactions during a decommit

v0d1ch opened the pull request:
fix #1526 

Related specification PR https://github.com/cardano-scaling/hydra-formal-specification/pull/7

### Why?

- New L2 transactions fail when decommit is _in flight_ because we try to re-apply the local decommit tx

### What

- Make sure the new snapshots are created only if the version is matched
- Any pending decommit needs to match with the decommit in confirmed snapshot thus preserved in the next snapshot/s until it is observed.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained herafter

ch1bo committed:
No processing of transactions during decommit

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -610,7 +620,7 @@ waitUntilMatch nodes predicate = do
       match seenMsgs n
 
-  oneMonth = 3600 * 24 * 30
+  oneMonth = 60 -- 3600 * 24 * 30
```
Must: not change this. I might have committed this wrongly, sorry.

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -472,18 +472,22 @@ onOpenNetworkReqSn env ledger st otherParty sv sn requestedTxIds mDecommitTx =
-            cont (activeUTxO, Just utxoToDecommit)
+    -- NOTE: if decommit tx was already applied when we constructed the previous
+    -- snapshot we should not try to apply it again
+    if confUTxOToDecommit == decommitUTxO
```
Although possible.. why do we re-request the `decommitTx` in this snapshot if it was already applied?

IMO it would be more intuitive to not even request it?

In any case, we **must** update the spec to match this. It's something we have overlooked in the algorithm.

v0d1ch replied:
I was thinking the same thing: Is it ok to just not include the decommit tx in the next snapshot if we already applied it? One question is how do we know if we applied it or not? we would have to keep some information about it in the local state I think. I would prefer if we don't populate the `utxoToDecommit` in the snapshot if it is not related to the snapshot in question. 

I think I need to draw this on the miro board to see what kind of scenarios are possible when you keep the decommit utxo around vs. when you omit it.

v0d1ch committed:
Introduce waiting on snapshot version in HeadLogic

On every ReqSn we wait to see the local version being the same as
the requested one before we request a new snapshot.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix red tests

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/json-schemas/logs.yaml:
```diff
@@ -2202,21 +2202,37 @@ definitions:
+          tag:
+            type: string
+            enum: ["WaitOnSnapshotVersion"]
+          waitingForNumber:
```
```suggestion
          waitingForVersion:
```

ch1bo commented on hydra-node/json-schemas/logs.yaml:
```diff
@@ -2202,21 +2202,37 @@ definitions:
+          waitingForNumber:
+            "$ref": "api.yaml#/components/schemas/SnapshotVersion"
+            description: >-
+              The expected number.
```
```suggestion
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -409,46 +409,48 @@ onOpenNetworkReqSn env ledger st otherParty sv sn requestedTxIds mDecommitTx =
-                    , utxo = u
-                    , utxoToDecommit = mUtxoToDecommit
+      -- Spec: wait v = v̂
+      waitOnSnapshotVersion $
```
Must: not `require` and `wait` for `v = v̂` at the same time

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -409,46 +409,48 @@ onOpenNetworkReqSn env ledger st otherParty sv sn requestedTxIds mDecommitTx =
+      waitOnSnapshotVersion $
+        -- Spec: require ̅S.𝑈 ◦ txω ≠ ⊥
+        --       ηω ← combine(outputs(txω))
+        --       𝑈_active ← ̅S.𝑈 ◦ txω \ outputs(txω)
```
Details are commented inside requireApplicableDecommitTx 

```suggestion
```

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -474,16 +482,37 @@ onOpenNetworkReqSn env ledger st otherParty sv sn requestedTxIds mDecommitTx =
+                    -- Spec: 𝑈_active ← ̅S.𝑈 ◦ txω \ outputs(txω)
+                    let utxoToDecommit = utxoFromTx decommitTx
+                    let activeUTxO = newConfirmedUTxO `withoutUTxO` utxoToDecommit
+                    cont (activeUTxO, Just utxoToDecommit)
```
Could: align this closer to what we have in the spec, i.e. make this block work with `mDecommitTx`

That way, we would have the same conditionals. However, I think it is semantically equivalent to what we have in the spec (although hard to see).

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -448,7 +471,12 @@ spec = parallel $ do
+
+              waitMatch n2 $ \case
+                HeadIsContested{headId, snapshotNumber} -> guard $ headId == testHeadId && snapshotNumber == 1
+                _ -> Nothing
```
Why did we add this? Why is a contest expected here?

v0d1ch replied:
One node n2 issues a decommit and  n1 closes. n2 already managed to get the confirmed snapshot with something to decrement so I wanted to assert that contestation happened therefore it is ok to just fanout `utxoRef 1` instead of `utxoRefs [1,2]`

ch1bo replied:
Okay nice. Having this in a comment on the code would have been awesome :)

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -831,9 +861,8 @@ toOnChainTx now = \case
   FanoutTx{} ->
     OnFanoutTx{headId = testHeadId}
 
--- NOTE(SN): Deliberately long to emphasize that we run these tests in IOSim.
```
I think this still holds and we should keep it long (one hour is not long, this might rather be a couple days at some point)

v0d1ch replied:
I got the test timeouts because now contestation happens now in _can close with decommit in flight_ spec. Perhaps the deadline calculation is somehow wrong and causing too long contestations?

ch1bo commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -474,6 +473,35 @@ spec =
+            reqSn = receiveMessageFrom bob $ ReqSn 0 2 [] (Just decommitTx)
+
+        outcome <- runHeadLogic bobEnv ledger s0 $ do
+          step reqSn
```
Could: make this a multi-step test which does `ReqSn 0 1` and `AckSn` accordingly. That would be less "whiteboxy" and not depend so much on the internal state of `confirmedSnapshot` and `seenSnapshot`

ch1bo reviewed (changes requested):
Only bug I found was the require/wait mix

Could align it more with the spec though.

Also, the spec is now in another repo so we need two PRs. This PR then should point to the new, updated spec in its flake inputs.

ffakenz commented on hydra-node/json-schemas/logs.yaml:
```diff
@@ -2202,21 +2202,37 @@ definitions:
+          waitingForVersion:
+            "$ref": "api.yaml#/components/schemas/SnapshotVersion"
+            description: >-
+              The expected number.
```
```suggestion
              The expected version number.
```

v0d1ch replied:
@ffakenz description is duplicated here. I removed it as it was suggested by @ch1bo

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -418,13 +418,13 @@ spec = parallel $ do
                 openHead chain n1 n2
                 let decommitTx1 = SimpleTx 1 (utxoRef 1) (utxoRef 42)
-                send n2 (Decommit{decommitTx = decommitTx1})
+                send n1 (Decommit{decommitTx = decommitTx1})
```
why do we swap n1 for n2 in this spec?
should not it remain the same?

v0d1ch replied:
It is because n1 is actually the owner of `utxoRef 1` which is to be decommited here.

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -811,7 +841,7 @@ toOnChainTx now = \case
     OnDecrementTx
       { headId
-      , newVersion = version
+      , newVersion = version + 1
```
good catch!

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -433,10 +433,33 @@ spec = parallel $ do
 
+      it "can process transactions while decommit pending" $
+        shouldRunInSim $ do
+          -- NOTE: The simulated network has a block time of 20 (simulated) seconds.
```
i think this note should be where withSimulatedChainAndNetwork is defined.
wdyt?

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -433,10 +433,33 @@ spec = parallel $ do
+                let normalTx = SimpleTx 2 (utxoRef 2) (utxoRef 3)
+                send n2 (NewTx normalTx)
+                waitUntilMatch [n1, n2] $ \case
+                  SnapshotConfirmed{snapshot = Snapshot{confirmed}} -> 2 `elem` confirmed
```
would be great to assert the version hasn't been bumped here,
but it gets increased upon receiving DecommitFinalized

v0d1ch replied:
`DecommitFinalized` doesn't contain the version information. How would we compare the versions?

ffakenz replied:
don't we have a mean to know the state of the head?

do you think is out of scope to add the version info to `DecommitFinalized`?

v0d1ch replied:
We only have means of getting the Head UTxO afaik and adding the version to `DecommitFinalized` just to be able to assert something in test is not really adding much value to the users/head-operators, it's just something that could help us make sense but then I would add it to all of the messages if we deem it important.

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -710,7 +737,10 @@ simulatedChainAndNetwork initialChainState = do
-                      createAndYieldEvent nodes history localChainState $ toOnChainTx now tx
+                      -- Only observe "after one block"
+                      void . async $ do
+                        threadDelay blockTime
```
good one!

ffakenz commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -535,12 +563,11 @@ spec = parallel $ do
               withHydraNode aliceSk [] chain $ \n1 -> do
                 send n1 Init
                 waitUntil [n1] $ HeadIsInitializing testHeadId (fromList [alice])
-                simulateCommit chain (alice, utxoRef 1)
```
why?

v0d1ch replied:
it is not needed since we only look at the `HeadIsInitializing` logs

ffakenz commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -474,6 +473,35 @@ spec =
+        outcome <- runHeadLogic bobEnv ledger s0 $ do
+          step reqSn
+        outcome `shouldSatisfy` \case
+          Error RequireFailed{} -> True
```
i feel we are missing some explanation on why it failed as its unclear.

v0d1ch replied:
I matched on exact condition :+1:

v0d1ch committed:
PR review changes

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo reviewed (changes requested):
Must refer to an updated formal spec flake input. I'll review it again in more detail tomorrow

v0d1ch committed:
Update hydra-spec input

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -475,15 +480,27 @@ onOpenNetworkReqSn env ledger st otherParty sv sn requestedTxIds mDecommitTx =
+          then
+            if confUTxOToDecommit == Just (utxoFromTx decommitTx)
+              then cont (confirmedUTxO, confUTxOToDecommit)
+              else Error $ RequireFailed ReqSnDecommitNotSettled
```
:+1:

ch1bo reviewed (approved):
Tested this manually on a slow local devnet and can't reproduce the behavior from the bug - fixed! :muscle:

Side note though: If we have multiple snapshots with a `utxoToDecommit`, the `hydra-node` seems to submit it on each snapshot confirmation, resulting eventually to a race of multiple `decrementTx`s and errors reported like:

```
TxValidationErrorInCardanoMode (ShelleyTxValidationError ShelleyBasedEraBabbage (ApplyTxError (UtxowFailure (AlonzoInBabbageUtxowPredFailure (ExtraRedeemers [Alonz
oSpending (AsIx {unAsIx = 0})])) :| [UtxowFailure (AlonzoInBabbageUtxowPredFailure (PPViewHashesDontMatch (SJust (SafeHash \"10f0625aedbf7ada32774b7574fb8223808451d22806d3693f31b76cbbc6c3fa\")) (SJust (SafeHash \"5e13ca4c11ff1467512cde99f5e300f7dc19f39099fd83b8b3e35b80427914f8\")))),UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (ValueNotConservedUTxO (MaryValue (Coin 0) (MultiAsset (fromList []))) (MaryValue (Coin 86182523) (MultiAsset (fromList [(PolicyID {policyID = ScriptHash \"3bc4890827e9784f22ee115016d5dcaf073b407175398b04f8f23ae0\"},fromList [(\"4879647261486561645631\",1),(\"f8a68cd18e59a6ace848155a0e967af64f4d00cf8acee8adc95a6b0d\",1)])])))))),UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (BadInputsUTxO (fromList [TxIn (TxId {unTxId = SafeHash \"9762190e385ad32de7eafd14044748d98cb6cbfc9ea59ab4e11faa7fedf450c4\"}) (TxIx {unTxIx = 0}),TxIn (TxId {unTxId = SafeHash \"9762190e385ad32de7eafd14044748d98cb6cbfc9ea59ab4e11faa7fedf450c4\"}) (TxIx {unTxIx = 2})])))),UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure NoCollateralInputs)),UtxowFailure (UtxoFailure (AlonzoInBabbageUtxoPredFailure (InsufficientCollateral (DeltaCoin 0) (Coin 2586266))))])))
```
